File: src/Functional/OCaml/decorator.ml
type component = unit -> string

let base_component () = "alert"
let audit_decorator (component : component) () = "audit(" ^ component () ^ ")"
let encrypt_decorator (component : component) () = "enc(" ^ component () ^ ")"

let () =
  let base = base_component in
  let audited = audit_decorator base in
  let encrypted = encrypt_decorator base in
  let stacked = audit_decorator (encrypt_decorator base) in
  Printf.printf "base=%s\n" (base ());
  Printf.printf "audit=%s\n" (audited ());
  Printf.printf "encrypted=%s\n" (encrypted ());
  Printf.printf "stacked=%s\n" (stacked ())

File: src/Functional/Erlang/decorator.erl
-module(decorator).
-export([main/0]).

base_component() -> fun() -> "alert" end.
audit_decorator(Component) -> fun() -> "audit(" ++ Component() ++ ")" end.
encrypt_decorator(Component) -> fun() -> "enc(" ++ Component() ++ ")" end.

main() ->
    Base = base_component(),
    Audited = audit_decorator(Base),
    Encrypted = encrypt_decorator(Base),
    Stacked = audit_decorator(encrypt_decorator(Base)),
    io:format("base=~s~n", [Base()]),
    io:format("audit=~s~n", [Audited()]),
    io:format("encrypted=~s~n", [Encrypted()]),
    io:format("stacked=~s~n", [Stacked()]).
